geomesa-hbase/geomesa-hbase-spark/src/main/scala/org/locationtech/geomesa/spark/hbase/GeoMesaHBaseInputFormat.scala

```diff
@@ -0,0 +1,157 @@
+    private val nextFeature =
+      (filterOpt, transformSchema) match {
+        case (Some(filter), Some(ts)) =>
+          val indices = ts.getAttributeDescriptors.map { ad => sft.indexOf(ad.getLocalName) }
+          val fn = table.toFeaturesWithFilterTransform(sft, filter, Array.empty[TransformProcess.Definition], indices.toArray, ts)
+          nextFeatureFromOptional(fn)
+
+        case (Some(filter), None) =>
+          val fn = table.toFeaturesWithFilter(sft, filter)
+          nextFeatureFromOptional(fn)
+
+        case (None, Some(ts))         =>
+          val indices = ts.getAttributeDescriptors.map { ad => sft.indexOf(ad.getLocalName) }
+          val fn = table.toFeaturesWithTransform(sft, Array.empty[TransformProcess.Definition], indices.toArray, ts)
+          nextFeatureFromDirect(fn)
+
+        case (None, None)         =>
+          val fn = table.toFeaturesDirect(sft)
+          nextFeatureFromDirect(fn)
+      }
+
+    private val getId = table.getIdFromRow(sft)
+
+    override def initialize(split: InputSplit, context: TaskAttemptContext): Unit = {
+      reader.initialize(split, context)
+    }
+    override def getProgress: Float = reader.getProgress
+
+    override def nextKeyValue(): Boolean = nextKeyValueInternal()
+
+    /**
+      * Get the next key value from the underlying reader, incrementing the reader when required
+      */
+    private def nextKeyValueInternal(): Boolean = {
+      nextFeature()
+      if(staged != null) {
+        if (!hasId) {
+          val row = reader.getCurrentKey
+          val offset = row.getOffset
+          val length = row.getLength
+          staged.getIdentifier.asInstanceOf[FeatureIdImpl].setID(getId(row.get(), offset, length))
+        }
+      }
+
+      staged != null
+    }
+
+    override def getCurrentValue: SimpleFeature = staged
+
+    override def getCurrentKey = new Text(staged.getID)
+
+    override def close(): Unit = { reader.close() }
+  }
+
+  private def init(conf: Configuration) = if (sft == null) {
+    import scala.collection.JavaConversions._
+
+    val params = GeoMesaConfigurator.getDataStoreInParams(conf)
+    val ds = DataStoreFinder.getDataStore(new CaseInsensitiveMap(params).asInstanceOf[java.util.Map[_, _]]).asInstanceOf[HBaseDataStore]
+    sft = ds.getSchema(GeoMesaConfigurator.getFeatureType(conf))
+    val tableName = GeoMesaConfigurator.getTable(conf)
+    table = HBaseFeatureIndex.indices(sft, IndexMode.Read)
+      .find(t => t.getTableName(sft.getTypeName, ds) == tableName)
+      .getOrElse(throw new RuntimeException(s"Couldn't find input table $tableName"))
+
+    delegate.setConf(conf)
+    // see TableMapReduceUtil.java
+    HBaseConfiguration.merge(conf, HBaseConfiguration.create(conf))
+    conf.set(TableInputFormat.INPUT_TABLE, tableName)
+  }
+
+
+  /**
+    * Gets splits for a job.
+    */
+  override def getSplits(context: JobContext): java.util.List[InputSplit] = {
+    init(context.getConfiguration)
```

Comment:
does this need to be called here, or could it be called from `initialize`? Not sure of the order things get called in...